mlir/test/Quantum/AdjointTest.mlir

```diff
@@ -0,0 +1,103 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// RUN: quantum-opt --adjoint-lowering --split-input-file %s | FileCheck %s
+
+// CHECK:      workflow_plain
+func.func private @workflow_plain() -> tensor<4xcomplex<f64>> attributes {} {
+  %c1_i64 = arith.constant 1 : i64
+  %cst = arith.constant 4.000000e-01 : f64
+  %c0_i64 = arith.constant 0 : i64
+  quantum.device ["backend", "lightning.qubit"]
+  %0 = quantum.alloc( 2) : !quantum.reg
+  %1 = quantum.extract %0[%c0_i64] : !quantum.reg -> !quantum.bit
+  // CHECK:        RX
+  %2 = quantum.custom "RX"(%cst) %1 : !quantum.bit
+  %3 = quantum.insert %0[%c0_i64], %2 : !quantum.reg, !quantum.bit
+  %4 = quantum.adjoint(%3) : (!quantum.reg) -> !quantum.reg {
+  // CHECK:        PauliZ
+  // CHECK-SAME:          adjoint = true
+  // CHECK:        PauliY
+  // CHECK-SAME:          adjoint = true
+  // CHECK:        PauliX
+  // CHECK-SAME:          adjoint = true
+  ^bb0(%arg0: !quantum.reg):
+    %10 = quantum.extract %arg0[%c0_i64] : !quantum.reg -> !quantum.bit
+    %11 = quantum.custom "PauliX"() %10 : !quantum.bit
+    %12 = quantum.custom "PauliY"() %11 : !quantum.bit
+    %13 = quantum.insert %arg0[%c0_i64], %12 : !quantum.reg, !quantum.bit
+    %15 = quantum.extract %13[%c1_i64] : !quantum.reg -> !quantum.bit
+    %16 = quantum.custom "PauliZ"() %15 : !quantum.bit
+    %17 = quantum.insert %13[%c1_i64], %16 : !quantum.reg, !quantum.bit
+    quantum.yield %17 : !quantum.reg
+  }
+  %5 = quantum.extract %4[%c0_i64] : !quantum.reg -> !quantum.bit
+  // CHECK:        RY
+  %6 = quantum.custom "RY"(%cst) %5 : !quantum.bit
+  %7 = quantum.extract %4[%c1_i64] : !quantum.reg -> !quantum.bit
+  %8 = quantum.compbasis %6, %7 : !quantum.obs
+  %9 = quantum.state %8 : tensor<4xcomplex<f64>>
+  quantum.dealloc %0 : !quantum.reg
+  return %9 : tensor<4xcomplex<f64>>
+}
+
+
+// CHECK:      workflow_nested
+// CHECK:      OpC
+// CHECK:      OpD
+// CHECK:      OpF
+// CHECK-SAME:      adjoint = true
+// CHECK:      OpE
+// CHECK-SAME:      adjoint = true
+// CHECK:      OpB
+// CHECK-SAME:      adjoint = true
+// CHECK:      OpA
+// CHECK-SAME:      adjoint = true
+func.func private @workflow_nested() -> tensor<4xcomplex<f64>> attributes {} {
+  %c1_i64 = arith.constant 1 : i64
+  %c0_i64 = arith.constant 0 : i64
+  quantum.device ["backend", "lightning.qubit"]
+  %0 = quantum.alloc( 2) : !quantum.reg
+  %1 = quantum.adjoint(%0) : (!quantum.reg) -> !quantum.reg {
+  ^bb0(%arg0: !quantum.reg):
+    %6 = quantum.extract %arg0[%c1_i64] : !quantum.reg -> !quantum.bit
+    %7 = quantum.custom "OpA"() %6 : !quantum.bit
+    %8 = quantum.custom "OpB"() %7 : !quantum.bit
+    %9 = quantum.insert %arg0[%c1_i64], %8 : !quantum.reg, !quantum.bit
+    %10 = quantum.adjoint(%9) : (!quantum.reg) -> !quantum.reg {
+    ^bb0(%arg1: !quantum.reg):
```

Comment:
It would be nice if we could get this block to be implicit in the IR, similar to func, scf.for, and other ops of the sort.

Reply:
Make sense, but can we leave it for next PR? This would involve quite a number of corrections at all levels.

Reply:
@pengmai Do you know how that is implemented in the built in ops?

Reply:
It looks pretty involved. For instance, the `scf.for` op has a custom printer and parser because of the custom handling of how `iter_args` get passed to its region arguments. I would concur with @grwlf to not let it block this PR.

Reply:
Leaving this one as-is for now.